relengapi/blueprints/slaveloan/tasks.py

```diff
@@ -0,0 +1,199 @@
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            bound_task = None
+            loanid = kwargs.get("loanid", None)
+            if args and isinstance(args[0], celery.Task):
+                bound_task = args[0]
+            if before:
+                add_task_to_history(loanid, before.format(**locals()))
+            retval = f(*args, **kwargs)
+            if after:
+                add_task_to_history(loanid, after.format(**locals()))
+            return retval
+        return wrapper
+    return decorator
+
+
+@task(bind=True)
+@add_to_history(
+    before="Choosing an inhouse machine based on slavealloc",
+    after="Chose inhouse machine {retval!s}")
+def choose_inhouse_machine(self, loanid, loan_class):
+    logger.debug("Choosing inhouse machine")
+    url = furl(current_app.config.get("SLAVEALLOC_URL", None))
+    # XXX: ToDo raise fatal if no slavealloc
+    url.path.add("slaves")
+    url.args["enabled"] = 1
+    try:
+        all_slaves = requests.get(str(url)).json()
+    except RequestException as exc:
+        logger.exception("Exception: %s" % exc)
+        self.retry(exc=exc)
+    # pylint silence
+    # available_slaves = filter(slave_mappings.slave_filter(loan_class), all_slaves)
+    available_slaves = [slave for slave in all_slaves
+                        if slave_mappings.slave_filter(loan_class)(slave)]
+    chosen = random.choice(available_slaves)
+    logger.debug("Chosen Slave = %s" % chosen)
+    return chosen['name']
+
+
+@task(bind=True, max_retries=None)
+@add_to_history(
+    before="Identifying FQDN and IP of {args[1]}",
+    after="Aquired FQDN and IP")
+def fixup_machine(self, machine, loanid):
+    try:
+        fqdn = socket.getfqdn("%s.build.mozilla.org" % machine)
+        ipaddr = socket.gethostbyname("%s.build.mozilla.org" % machine)
+        session = current_app.db.session('relengapi')
+        m = Machines.as_unique(session,
+                               fqdn=fqdn,
+                               ipaddr=ipaddr)
+        l = session.query(Loans).get(loanid)
+        l.machine = m
+        session.commit()
+        l = session.query(Loans).get(loanid)
+    except Exception as exc:  # pylint: disable=W0703
+        logger.exception(exc)
+        self.retry(exc=exc)
+
+
+@task(bind=True)
+@add_to_history(
+    before="Setup tracking bug for {args[1]}",
+    after="Tracking bug {retval!s} linked with loan")
+def bmo_set_tracking_bug(self, machine, loanid):
+    try:
+        session = current_app.db.session('relengapi')
+        l = session.query(Loans).get(loanid)
+        assert l.bug_id
+
+        bug_comment = "Being loaned to %s in Bug %s" % (l.human.ldap, l.bug_id)
+
+        tracking_bug = bugzilla.ProblemTrackingBug(machine, loadInfo=False)
+        try:
+            tracking_bug.refresh()
+        except bzrest.errors.BugNotFound:
+            logger.info("Couldn't find bug, creating it...")
+            tracking_bug.create(comment=bug_comment, depends_on=l.bug_id)
```

Comment:
It seems like this update-or-create functionality should be within the ProblemTrackingBug class.